### PR TITLE
Keep soroban_state Arc-count perf fields non-cumulative in LedgerClosePerf aggregation

### LINKED ISSUE #2823: LEDGER: Keep soroban_state Arc-count perf fields non-cumulative in LedgerClosePerf aggregation
Deferred from PR #2812 at the review lifetime cap.

## Residual concern

`LedgerClosePerf::AddAssign` currently sums the new `soroban_state_data_arc_count` and `soroban_state_code_arc_count` fields in `crates/ledger/src/close.rs`.

These fields are instantaneous per-close observations, not cumulative counters. Summing them produces meaningless aggregate values for downstream consumers that fold `LedgerClosePerf` across ledgers, such as `crates/simulation/src/applyload.rs`.

## Expected follow-up

- Change the aggregation semantics for these two fields from sum to latest-value (or otherwise keep them out of aggregated perf output).
- Update the corresponding unit test so it no longer blesses cumulative behavior.
- Re-check any downstream consumer that uses `agg += perf` to ensure the new semantics are documented and intentional.

## References

- Parent issue: #2810
- Deferred from PR: #2812
- Reviewer note: `crates/ledger/src/close.rs:1096-1097`, `crates/simulation/src/applyload.rs:1128-1150`

### PR DESCRIPTION
**Pull request by tomerweller**
Closes #2823

## Summary

Changes `LedgerClosePerf::AddAssign` so that `soroban_state_data_arc_count` and `soroban_state_code_arc_count` are assigned from `rhs` (latest-sample semantics) rather than summed. These fields are instantaneous per-close observations, not cumulative counters, so summing them across ledgers produces meaningless values.

## Plan reference

[Converged Plan comment](https://github.com/stellar-experimental/henyey/issues/2823#issuecomment-4500217554)

## Test plan

- [x] cargo fmt --check
- [x] cargo clippy -p henyey-ledger -- -D warnings
- [x] cargo test -p henyey-ledger passes

## Regression test

- **Test:** `crates/ledger/src/close.rs::test_ledger_close_perf_add_assign_keeps_soroban_arc_counts_latest`
- **Pre-fix:** committed as `72252dcb` — verified FAILED with: `assertion left == right failed: left: 5, right: 3`
- **Post-fix:** verified PASSES after `b3e831af`

## Deviations from plan

- `crates/simulation/src/applyload.rs` does not reference the arc-count fields at all, so no comment was needed there.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** APPROVE

**Summary:** The PR changes only benchmark/perf aggregation semantics for Henyey-only Soroban Arc-count diagnostics.
It does not alter ledger-close sequencing or results, and the switch to latest-sample semantics matches stellar-core's use of per-close/current-state gauges rather than cumulative counters.

<details>
<summary>Full review</summary>

- `soroban_state_data_arc_count` / `soroban_state_code_arc_count` are Henyey-only diagnostics sampled before `make_mut` in the Soroban state update path (`crates/ledger/src/manager.rs:5222-5227`); I found no corresponding Arc-strong-count field in `stellar-core/`, so there is no upstream consensus/output behavior to preserve here.
- The changed code is in `LedgerClosePerf::add_assign`, and the only non-test callsite is benchmark aggregation in `crates/simulation/src/applyload.rs:1128-1149`. That means this PR affects reporting/benchmark summaries, not ledger-close sequencing, tx application order, or emitted ledger results.
- For the closest stellar-core analogue, upstream treats comparable per-ledger / current-state observability fields as gauges set to the latest value, not cumulative sums: `stellar-core/src/ledger/LedgerManagerImpl.cpp:2605-2615` updates stage/cluster metrics with `set_count(...)`, `stellar-core/src/ledger/LedgerManager.h:293-295` describes these as “current state” gauge-like counters, and `stellar-core/src/ledger/InMemorySorobanState.cpp:624-629` likewise publishes in-memory Soroban state metrics with `set_count(...)`.
- On that basis, replacing `+=` with assignment for these instantaneous Arc-count samples is parity-aligned with stellar-core's metrics model and removes a Henyey-only reporting artifact without introducing any parity divergence.

</details>

File: crates/ledger/src/close.rs [hunk 1]
```diff
@@ -1093,10 +1093,11 @@ impl std::ops::AddAssign<&LedgerClosePerf> for LedgerClosePerf {
...
         self.post_exec_us += rhs.post_exec_us;
         self.tx_count += rhs.tx_count;
-        self.soroban_state_data_arc_count += rhs.soroban_state_data_arc_count;
-        self.soroban_state_code_arc_count += rhs.soroban_state_code_arc_count;
-        // soroban_stage_count and soroban_max_cluster_count are per-close
+        // soroban_state_*_arc_count, soroban_stage_count, and
+        // soroban_max_cluster_count are per-close instantaneous observations
         // (not cumulative), so take the latest rather than summing.
+        self.soroban_state_data_arc_count = rhs.soroban_state_data_arc_count;
+        self.soroban_state_code_arc_count = rhs.soroban_state_code_arc_count;
```
[the remark this answers is not in the document]
Comment (tomerweller) [reply crates/ledger/src/close.rs:1100]: Addressed in d61902d5: now assigns both `soroban_stage_count` and `soroban_max_cluster_count` from `rhs` in `add_assign`, and extends the regression test to cover them.

**Review by tomerweller — COMMENTED** [review]
Addressed feedback. Ready for re-review.

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** APPROVE

**Summary:** The updated aggregation semantics are parity-safe: these fields are observational perf data, not ledger state, and the follow-up now covers every per-close field in this block that should be latest-sample rather than cumulative.
The closest stellar-core analogue (`ledger.apply-soroban.{stages,max-clusters}`) is also exposed as per-close gauge state via `set_count(...)`, not as a running sum.

<details>
<summary>Full review</summary>

- `soroban_state_data_arc_count` / `soroban_state_code_arc_count` are Henyey-only diagnostics sampled immediately before the Soroban state `make_mut` path (`crates/ledger/src/manager.rs:5222-5227`). I did not find a stellar-core field that exposes Arc strong counts on this path, so there is no upstream consensus/output value to preserve; parity here is about not inventing misleading observability semantics.
- On that basis, switching these Arc-count fields from `+=` to assignment in `LedgerClosePerf::add_assign` is the parity-aligned choice. A cumulative sum would fabricate a quantity stellar-core never reports, while assignment preserves the last observed per-close sample.
- The follow-up commit was necessary and correct: `soroban_stage_count` and `soroban_max_cluster_count` are documented in `LedgerClosePerf` as per-close structure (`0 if no Soroban phase`), so leaving them out of the latest-sample block would have kept the aggregation semantics internally inconsistent.
- For the closest upstream counterpart, stellar-core records Soroban parallel structure as gauge-like current/per-close state, not cumulative totals: `stellar-core/src/ledger/LedgerManagerImpl.cpp:2605-2615` calls `set_count(maxClusters)` and `set_count(applyStages.size())`, and `stellar-core/src/ledger/LedgerManager.h:293-295` explicitly describes these metrics as “current state” gauge-like counters.
- The simulation benchmark path also matches that reading: `stellar-core/src/simulation/ApplyLoad.cpp:1719-1728` inspects the current `ledger.apply-soroban.stages` / `max-clusters` counters after a close, rather than aggregating them across closes.
- Edge case check: after this patch, aggregating a mixed run that ends with a non-Soroban ledger will leave `soroban_stage_count == 0` and `soroban_max_cluster_count == 0`. That matches the field definitions in `LedgerClosePerf` itself (“0 if no Soroban phase”). Henyey's app-level exported metrics remain separately sticky at `crates/app/src/app/ledger_close.rs:2228-2234`, so this change does not introduce a parity-significant sequencing mismatch in the live close path.
- I found no parity concern in ledger-close ordering, tx application, header/meta generation, or any other consensus-visible behavior: the only functional effect is on benchmark/report aggregation (`crates/simulation/src/applyload.rs:1128-1149`).

</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** APPROVE

**Summary:** Verified the bug-fix workflow end-to-end: issue #2823 is `kind: bug-fix`, the PR body documents the regression test, and `close::tests::test_ledger_close_perf_add_assign_keeps_soroban_arc_counts_latest` fails on `72252dcb` but passes on the final PR head.
The final diff is internally consistent: `LedgerClosePerf::add_assign` now keeps all per-close instantaneous fields in this block as latest-sample values, and the regression test covers both the arc-count and stage/max-cluster cases.

<details>
<summary>Full review</summary>

- Verified the linked issue's `## Triage Report` marks #2823 as `kind: bug-fix`, and the PR body includes the required `## Regression test` section with a matching test claim.
- Checked commit order in an isolated worktree at `~/data/22432486-c1b5-4144-affd-be4775ee3fcd/review-pr-2823-correctness`: `72252dcb` (test) → `b3e831af` (fix) → `d61902d5` (follow-up).
- Re-ran the claimed regression test exactly as `cargo test -p henyey-ledger close::tests::test_ledger_close_perf_add_assign_keeps_soroban_arc_counts_latest -- --exact`: it **failed** on `72252dcb` with `assertion \`left == right\` failed` (`left: 5`, `right: 3`), then **passed** on `d61902d5`.
- Reviewed the only behavior change in `crates/ledger/src/close.rs`: `LedgerClosePerf::add_assign` now assigns `soroban_state_data_arc_count`, `soroban_state_code_arc_count`, `soroban_stage_count`, and `soroban_max_cluster_count` from `rhs`, which matches their documented per-close / instantaneous semantics.
- Checked the consumer side in `crates/simulation/src/applyload.rs:1148-1149`: this aggregation is used for benchmark/reporting summaries (`agg += p`), so the fix changes observability output rather than ledger-close correctness, tx execution, or result generation.
- The follow-up commit was necessary and correct: it closes the earlier consistency gap where the comment claimed latest-sample semantics for stage/max-cluster fields without actually copying them, and the updated unit test now covers that behavior too.
- Alternative considered: narrowing the comment to mention only the two arc-count fields. The chosen implementation is better because `soroban_stage_count` / `soroban_max_cluster_count` are already documented as per-close observations in the same struct and should not remain stale or implicitly cumulative under aggregation.
- Readability and error handling look good for the scope here: the change is localized, the comment now matches the code, and no new fallible paths were introduced.
- No further correctness issues found.
</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** APPROVE

**Summary:** The diff fixes the parity issue on this code path by preserving latest-sample semantics for per-close Soroban perf observations instead of inventing cumulative values across ledgers.
This matches stellar-core's per-ledger handling of equivalent apply-soroban structure metrics.

<details>
<summary>Full review</summary>

- I inspected the full PR commit stack (`72252dcb` failing regression, `b3e831af` fix, `d61902d5` follow-up) and the final diff in `crates/ledger/src/close.rs`.
- In `crates/ledger/src/close.rs:1096-1102`, the changed `AddAssign` now assigns `rhs` for `soroban_state_{data,code}_arc_count`, `soroban_stage_count`, and `soroban_max_cluster_count` instead of summing them. For parity, this is the correct shape: these are instantaneous per-close observations, not ledger-to-ledger accumulators.
- The closest stellar-core reference is `stellar-core/src/ledger/LedgerManagerImpl.cpp:2605-2614`, where the corresponding per-ledger Soroban structure metrics (`ledger.apply-soroban.stages` and `ledger.apply-soroban.max-clusters`) are written with `set_count(...)`, not incremented. Stellar-core therefore preserves the latest ledger's sample rather than accumulating across closes.
- The arc-count fields are henyey-only instrumentation, but they are sampled at a single point in the close path and belong to the same "latest sample" category. Summing them across closes would create values with no stellar-core analogue and would misrepresent the per-ledger observation.
- The regression test in `crates/ledger/src/close.rs:2008-2041` now covers both arc-count fields and the stage/max-cluster fields, which is the relevant edge on this changed aggregation path.
- I did not find a remaining parity divergence in the patched code path relative to stellar-core semantics.
</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** APPROVE

**Summary:** Verified the bug-fix workflow end-to-end: the PR body's regression-test section matches the diff, the named test fails on the pre-fix test commit, and passes on the final PR head.
The final diff is internally consistent: `LedgerClosePerf::add_assign` now keeps all per-close instantaneous fields in this block as latest-sample values, and the regression test covers both the arc-count and stage/max-cluster cases.

<details>
<summary>Full review</summary>

- Confirmed issue #2823 is triaged as **Kind: bug-fix** with a regression-test obligation, and the PR body includes a `## Regression test` section naming `crates/ledger/src/close.rs::test_ledger_close_perf_add_assign_keeps_soroban_arc_counts_latest` plus the claimed pre-fix/post-fix commits.
- Verified the PR history matches that claim: `72252dcb` adds the regression test, `b3e831af` fixes the two arc-count fields, and `d61902d5` follows up by making `soroban_stage_count` / `soroban_max_cluster_count` obey the same documented latest-sample semantics.
- Re-ran the named regression test exactly as `cargo test -p henyey-ledger close::tests::test_ledger_close_perf_add_assign_keeps_soroban_arc_counts_latest -- --exact` in an isolated worktree: it **failed** on `72252dcb` with `assertion \`left == right\` failed` (`left: 5`, `right: 3`), so the test would have caught the bug.
- Re-ran the same focused test on the final PR head `d61902d5`; it **passed**, confirming the final patch repairs the reported behavior.
- Reviewed the only code change in `crates/ledger/src/close.rs`: assigning `soroban_state_data_arc_count`, `soroban_state_code_arc_count`, `soroban_stage_count`, and `soroban_max_cluster_count` from `rhs` is consistent with the struct docs marking them as per-close / instantaneous observations rather than cumulative counters.
- Checked the producer/consumer context: the arc counts are sampled once per close before `Arc::make_mut` in `crates/ledger/src/manager.rs`, and the in-tree fold site in `crates/simulation/src/applyload.rs` uses `agg += p` only for benchmark/report aggregation. I found no correctness impact on ledger-close sequencing, transaction execution, or result generation.
- The follow-up change for `soroban_stage_count` / `soroban_max_cluster_count` was necessary for internal consistency, and the expanded regression test now covers that adjacent correctness case as well.
- No additional correctness blockers found in this diff.
</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** APPROVE

**Summary:** This is parity-safe: it only changes henyey-side aggregation of henyey-only perf samples, and it corrects instantaneous fields to use latest-sample semantics instead of cross-ledger accumulation.

<details>
<summary>Full review</summary>

- The touched code is `LedgerClosePerf::add_assign` in `crates/ledger/src/close.rs`, and its aggregation use here is henyey-side only (`crates/simulation/src/applyload.rs:1149`); it does not affect transaction application, ledger header construction, result ordering, or meta/XDR emission.
- `soroban_state_data_arc_count` / `soroban_state_code_arc_count` are henyey-only observability fields (`henyey_ledger_close_soroban_state_data_arc_refs` and `henyey_ledger_close_soroban_state_code_arc_refs` in `crates/app/src/metrics.rs:805-807`). I did not find a 1:1 stellar-core analogue for Arc `strong_count` sampling in the ledger-close path.
- The nearest stellar-core behavior points the same way semantically: `stellar-core/src/ledger/InMemorySorobanState.cpp:626-629` publishes in-memory state metrics with `set_count(...)` (absolute/instantaneous values, not cumulative cross-ledger sums), and `stellar-core/src/ledger/SorobanMetrics.cpp:206-224` publishes ledger-wide counters and then resets them per ledger.
- Extending the same latest-sample treatment to `soroban_stage_count` and `soroban_max_cluster_count` is also parity-safe: these are per-close shape observations, so replacing accumulation with overwrite avoids a misleading multi-ledger total without changing any protocol-visible behavior.
- I found no divergence in close sequencing, edge-case behavior, or stellar-core-visible semantics from this patch.
</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** APPROVE

**Summary:** The regression test in the PR body matches the diff and does fail on commit 72252dcb; the current head fixes the aggregation bug by keeping the per-close soroban Arc-count fields non-cumulative.

<details>
<summary>Full review</summary>

- Verified the PR body’s regression-test section matches the code: `test_ledger_close_perf_add_assign_keeps_soroban_arc_counts_latest` is the added unit test in `crates/ledger/src/close.rs`.
- Checked out commit `72252dcbc7fe` in an isolated worktree and ran `cargo test -p henyey-ledger test_ledger_close_perf_add_assign_keeps_soroban_arc_counts_latest`; it failed with `assertion left == right failed` (`left: 5`, `right: 3`), confirming the test reproduces the pre-fix bug.
- Checked out the current PR head and reran the same test; it passed.
- Reviewed `LedgerClosePerf::AddAssign`: the two soroban Arc-count fields now correctly use latest-sample semantics instead of cumulative summation, which matches their documented meaning as instantaneous per-close observations.
- The accompanying test coverage is adequate and also exercises the already-documented latest-value semantics for `soroban_stage_count` and `soroban_max_cluster_count`.
</details>
